### PR TITLE
Dev apply transformer

### PR DESCRIPTION
**Pull request by taraspiotr**
- added option to cache step output in memory and caching mask_resize output
- create make_apply_trasformer function in utils.py that creates transformer applying given function over input
- removed redundant transformers and replaced them with make_apply_trasformer
- added docstrings for functions in postprocessing.py
- removed stream mode
- removed CRF from postprocessing pipeline

### CODE REVIEW

File: src/models.py [hunk 1]
```diff
@@ -42,7 +42,16 @@
 
 
 class BasePyTorchUNet(Model):
+    """Base class for PyTorch U-net models.
+
+    Args:
```
Comment (jakubczakon) [inline src/models.py]: @taraspiotr todo?

File: src/callbacks.py [hunk 2]
```diff
@@ -200,29 +200,35 @@ def _generate_prediction(self, cache_dirpath, outputs):
 
 def postprocessing__pipeline_simplified(cache_dirpath):
     mask_resize = Step(name='mask_resize',
-                       transformer=post.Resizer(),
+                       transformer=make_apply_transformer(post.resize_image,
```
Comment (jakubczakon) [inline src/callbacks.py]: @taraspiotr I like how nice and easy to use it is.

File: src/steps/base.py [hunk 3]
```diff
@@ -56,6 +56,7 @@ def _prep_cache(self, cache_dirpath):
         self.cache_filepath_step_transformer = os.path.join(self.cache_dirpath_transformers, self.name)
         self.save_filepath_step_output = os.path.join(self.save_dirpath_outputs, '{}'.format(self.name))
         self.save_filepath_step_tmp = os.path.join(self.save_dirpath_tmp, '{}'.format(self.name))
+        self._cached_output = None
```
Comment (jakubczakon) [inline src/steps/base.py]: @taraspiotr so now we don't have this load/save cache option available at all. It is ok for this project but for other projects (general use) I'd rather have that available so when you make a PR to steppy make sure to deal with both options. That suggest it is probably a good spot to deal with both here not to have vastly different codebase.

File: src/utils.py [hunk 4]
```diff
@@ -411,3 +408,37 @@ def add_dropped_objects(original, processed):
         if not np.any(np.where((labeled == i) & processed)):
             reconstructed += (labeled == i)
     return reconstructed.astype('uint8')
+
+
+def make_apply_transformer(func, output_name='output', apply_on=None):
+    class StaticApplyTransformer(BaseTransformer):
+        def transform(self, *args, **kwargs):
+            iterator_length = self.check_input(*args, **kwargs)
```
Comment (jakubczakon) [inline src/utils.py]: @taraspiotr I don't like the naming. check_input can return True or do some assertion but I don't like that it returns length. I would rather have 2 methods `check_input` and `get_length` that we run one ofter the other,

**Comment by jakubczakon** [comment]
@taraspiotr I think CRF is still left in the `neptune.yaml`

File: src/models.py (hunk 1 above)
Comment (taraspiotr) [reply src/models.py]: @jakubczakon my bad, I've just started documenting models, but in this PR there should be only postprocessing docstrings

File: src/steps/base.py (hunk 3 above)
Comment (taraspiotr) [reply src/steps/base.py]: @jakubczakon we still have an option to save output on disk, but yes, cache in memory is implemented in place of cache on disk, with the same interface. Right know in steppy cache and save are pretty much the same with the only difference being that cache is saved in temp directory. I think it is worth rethinking if caching on disk alongside saving on disk is not a bit redundant.

File: src/utils.py (hunk 4 above)
Comment (taraspiotr) [reply src/utils.py]: @jakubczakon I fully agree, at first, this function was just checking input and then I added length on the spot without rethinking naming.

File: src/steps/base.py (hunk 3 above)
Comment (jakubczakon) [reply src/steps/base.py]: @taraspiotr we've discussed and figured that the difference between cache and save is that cache is during one run (cleaned after) saved is done between runs (no cleaning). If you do decide to save you need to load it on purpuse as well. You can also load those pickle objects somewhere else (notebook) to investigate. So I think we potentially all of the cases dealt with (cache on disk, cache in memory, save on disk).

I will merge it here but PR for steppy would need to be a bit more general.

Comment (taraspiotr) [reply src/steps/base.py]: @jakubczakon ok, sure

**Pull request merged**